src/utils/allocators.zig

```diff
@@ -178,120 +192,154 @@ pub const DiskMemoryAllocator = struct {
-                return null;
-            };
-            file.seekTo(0) catch |err| {
-                std.debug.print("Disk Memory Allocator error: {}\n", .{err});
-                return null;
-            };
-        }
+        };
 
-        const memory = std.posix.mmap(
+        const full_alloc = std.posix.mmap(
             null,
             aligned_size,
             std.posix.PROT.READ | std.posix.PROT.WRITE,
             std.posix.MAP{ .TYPE = .SHARED },
             file.handle,
             0,
         ) catch |err| {
-            std.debug.print("Disk Memory Allocator error: {}\n", .{err});
+            self.logFailure(err, file_name);
             return null;
         };
 
-        return memory.ptr;
-    }
-
-    /// unmaps the memory (file still exists and is removed on deinit())
-    pub fn free(_: *anyopaque, buf: []u8, log2_align: u8, return_address: usize) void {
-        _ = log2_align;
-        _ = return_address;
-        // TODO: build a mapping from ptr to file so we can delete the corresponding file on free
-        const buf_aligned_len = std.mem.alignForward(usize, buf.len, std.mem.page_size);
-        std.posix.munmap(@alignCast(buf.ptr[0..buf_aligned_len]));
+        std.mem.bytesAsValue(Metadata, full_alloc[size..][0..@sizeOf(Metadata)]).* = .{
+            .file_index = file_index,
+        };
+        return full_alloc.ptr;
     }
 
     /// not supported rn
     fn resize(
-        _: *anyopaque,
-        buf_unaligned: []u8,
-        log2_buf_align: u8,
+        ctx: *anyopaque,
+        buf: []u8,
+        log2_align: u8,
         new_size: usize,
         return_address: usize,
     ) bool {
-        // not supported
-        _ = buf_unaligned;
-        _ = log2_buf_align;
-        _ = new_size;
         _ = return_address;
-        return false;
+        const self: *Self = @ptrCast(@alignCast(ctx));
+
+        const alignment = @as(usize, 1) << @intCast(log2_align);
+        std.debug.assert(alignment <= std.mem.page_size); // the allocator interface shouldn't allow this (aside from the *Raw methods).
+
+        const buf_ptr: [*]align(std.mem.page_size) u8 = @alignCast(buf.ptr);
+        const aligned_size = fullMmapSize(buf.len);
+        const new_aligned_size = fullMmapSize(new_size);
+
+        const metadata: Metadata = @bitCast(buf_ptr[buf.len..][0..@sizeOf(Metadata)].*);
+        const file_name_bounded = fileNameBounded(metadata.file_index);
+        const file_name = file_name_bounded.constSlice();
+
+        if (aligned_size == new_aligned_size) return true;
+
+        if (new_aligned_size < aligned_size) {
+            std.posix.munmap(@alignCast(buf_ptr[new_aligned_size..aligned_size]));
+            std.mem.bytesAsValue(Metadata, buf_ptr[new_size..][0..@sizeOf(Metadata)]).* = .{
+                .file_index = metadata.file_index,
+            };
+            return true;
+        } else {
+            const file = self.dir.openFile(file_name, .{ .mode = .read_write }) catch |err| {
```

Comment:
Doesn't the file need to be resized? Won't you get a SIGBUS when you try to access mmapped memory that is out of the file bounds?

> SIGBUS   Attempted access to a page of the buffer that lies beyond the end of the mapped file.

https://man7.org/linux/man-pages/man2/mmap.2.html

Reply:
* Fixed the bug: 9afa8d5e369f503a9716c34dd3813bbb88af27bc
* Improvement: c3dc473a41c1d900ff434a7a4e3c9c375f386b25